go/plugins/postgresql/docstore.go

```diff
@@ -0,0 +1,157 @@
+package postgresql
+
+import (
+	"context"
+	"fmt"
+	"slices"
+
+	"github.com/firebase/genkit/go/ai"
+)
+
+type docStore struct {
+	engine PostgresEngine
+
+	tableName             string
+	schemaName            string
+	idColumn              string
+	metadataJsonColumn    string
+	contentColumn         string
+	embeddingColumn       string
+	metadataColumns       []string
+	ignoreMetadataColumns []string
+
+	embedder        ai.Embedder
+	embedderOptions any
+}
+
+func (ds *docStore) query(ctx context.Context, ss *SimilaritySearch, embbeding []float32) (*ai.RetrieverResponse, error) {
+	res := &ai.RetrieverResponse{}
+
+	query := ss.buildQuery(embbeding)
+	rows, err := ds.engine.Pool.Query(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	fieldDescriptions := rows.FieldDescriptions()
+	columnNames := make([]string, len(fieldDescriptions))
+
+	for i, fieldDescription := range fieldDescriptions {
+		columnNames[i] = fieldDescription.Name
+	}
+
+	for rows.Next() {
+		values := make([]interface{}, len(columnNames))
+		valuesPrt := make([]interface{}, len(columnNames))
+
+		for i := range columnNames {
+			valuesPrt[i] = &values[i]
+		}
+		if err := rows.Scan(valuesPrt...); err != nil {
+			return nil, fmt.Errorf("scan row failed: %v", err)
+		}
+
+		meta := make(map[string]any, ss.k)
+		var content []*ai.Part
+		for i, col := range columnNames {
+			if (len(ss.metadataColumns) > 0 && !slices.Contains(ss.metadataColumns, col)) &&
+				ss.contentColumn != col &&
+				ss.metadataJsonColumn != col {
+				continue
+			}
+
+			if ss.contentColumn == col {
+				content = append(content, ai.NewTextPart(values[i].(string)))
+			}
+
+			if ss.metadataJsonColumn == col {
+				content = append(content, ai.NewJSONPart(values[i].(string)))
+				continue
+			}
+
+			meta[col] = values[i]
+		}
+
+		doc := &ai.Document{
+			Metadata: meta,
+			Content:  content,
+		}
+
+		res.Documents = append(res.Documents, doc)
+	}
+
+	return res, nil
+}
+
+type RetrieverOptions struct {
+	SimilaritySearch *SimilaritySearch
+}
+
+func (ds *docStore) Retrieve(ctx context.Context, req *ai.RetrieverRequest) (*ai.RetrieverResponse, error) {
+	if req.Options == nil {
+		ss := NewSimilaritySearch()
```

Comment:
Just add the similarity search logic here instead of using the separate struct and methods